jupyterhub_singleuser_profiles/service.py

```diff
@@ -37,91 +40,114 @@ def __init__(self, server_url, token, namespace=None, verify_ssl=True):
     self.k8s_api_instance = client.CoreV1Api(k8s_client)
     self.os_client = DynamicClient(k8s_client)
 
-  def get_template(self, name):
-    tmplt = self.os_client.resources.get(api_version='template.openshift.io/v1', kind='Template')
-    if isinstance(tmplt, list):
-      tmplt = tmplt[0] # Openshift returns list of 2 resource types, we want template.openshift.io/v1/processedtemplates
+  def get_service_reference_config_map (self, user):
+    config_map_wrapper = self.os_client.resources.get(api_version='v1', kind='ConfigMap')
+    try:
+      body = {
+          'kind': 'ConfigMap',
+          'apiVersion': 'v1',
+          'metadata': {'name': _REFERENCE_CM_NAME %(escape(user))},
+          'data': {}
+        }
+      config_map_wrapper.create(body=body, namespace=self.namespace)
+
+    except ApiException as e:
+      if e.status != 409:
+        _LOGGER.error("Error getting reference ConfigMap in %s for %s: %s\n" % (self.namespace, escape(user), e))
+        raise
+
+    result = config_map_wrapper.get(namespace=self.namespace, name=_REFERENCE_CM_NAME %(escape(user)))
+    return result
 
+
+  def get_template(self, name, path):
+    cm_wrapper = self.os_client.resources.get(api_version='v1', kind='ConfigMap')
     try:
-      response = tmplt.get(
+      response = cm_wrapper.get(
           namespace=self.namespace,
           name=name
       )
-      template = response.to_dict()
+      cm = response.to_dict()
     except Exception as e:
       _LOGGER.error("Error: %s %s" % (name, e))
       template = None
+    template = yaml.load(cm['data'].get(path))
 
     return template
 
-  def process_template(self, user, template, **parameters):
-
-    self._set_template_parameters(
-      template,
-      USERNAME=escape(user),
-      **parameters
-    )
-    template["apiVersion"] = "template.openshift.io/v1"
-
-    endpoint = "{}/apis/template.openshift.io/v1/namespaces/{}/processedtemplates".format(
-        self.server_url,
-        self.namespace
-    )
-    response = requests.post(
-        endpoint,
-        json=template,
-        headers={
-            'Authorization': 'Bearer {}'.format(self.token),
-            'Content-Type': 'application/json'
-        },
-        verify=False #FIXME?
-    )
-    _LOGGER.debug("OpenShift master response template (%d): %r", response.status_code, response.text)
+  def process_template(self, user, service_name, template, configuration, labels=None):
+
+    tmp = jinja2.Template(json.dumps(template))
+
+    result = tmp.render(configuration)
 
+    result = json.loads(result)
+    if not result.get('metadata'):
+      result['metadata'] = {}
+    if labels:
+      result['metadata']['labels'].update(labels)
+    
+    if (not configuration.get('user')):
+      result['metadata']['name'] = service_name + '-' + user
+
+    return result
+
+  def add_owner_reference(self, processed_template, user):
```

Comment:
Can we change this to `get_owner_reference` and make it so that it only returns the ownerReference object

i.e.
```
     {
            'kind' : ref_cm['kind'],
            'apiVersion' : ref_cm['apiVersion'],
            'name' : ref_cm['metadata']['name'],
            'uid' : ref_cm['metadata']['uid']
          }
```


That way you only need to call it once at the beginning of `get_template_per_service` and we save some API calls (and time) since the CM is only created once and does not change if there are multiple services or resources